### PR TITLE
[SPARK-41472][CONNECT][PYTHON] Implement the rest of string/binary functions

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Implement the rest of string/binary functions. The first commit is https://github.com/apache/spark/pull/38921.

Among them, `format_number` has data type mismatch issue and should be enabled with https://issues.apache.org/jira/browse/SPARK-41473.

### Why are the changes needed?
For API coverage on Spark Connect.

### Does this PR introduce _any_ user-facing change?
Yes. New functions are available on Spark Connect.

### How was this patch tested?
Unit tests.

### CODE REVIEW

**Comment by zhengruifeng** [comment]
LGTM pending python linter

**Review by zhengruifeng — APPROVED** [review]

File: python/pyspark/sql/connect/functions.py
Comment (zhengruifeng) [inline python/pyspark/sql/connect/functions.py:4162]: I will resolve it in a batch, after this PR get merged

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/connect/functions.py
Comment (zhengruifeng) [inline python/pyspark/sql/connect/functions.py:5729]:
```suggestion
    timestamp: Optional["ColumnOrName"] = None, format: str = "yyyy-MM-dd HH:mm:ss"
```

**Comment by zhengruifeng** [comment]
Let me take over this PR in https://github.com/apache/spark/pull/39071

**Pull request closed without merging**
